src/filters/filters.jl

```diff
@@ -1,33 +1,163 @@
+* `register_filter` - Defaults to using the above functions to register the filter
+
+Implement the Advanced Methods to avoid @cfunction from generating a runtime closure which may not work on all systems.
+"""
 abstract type Filter end
 
 """
     FILTERS
 
 Maps filter id to filter type.
 """
-const FILTERS = Dict{API.H5Z_filter_t, Any}()
+const FILTERS = Dict{API.H5Z_filter_t, Type{<: Filter}}()
 
 """
-    filterid(::F)
-    filterid(F)
+    filterid(::F) where {F <: Filter}
+    filterid(F) where {F <: Filter}
+
 
 The internal filter id of a filter of type `F`.
 """
 filterid(::F) where {F<:Filter} = filterid(F)
 
+"""
+    encoder_present(::F) where {F<:Filter}
+
+Can the filter have an encode or compress the data?
+Defaults to true.
+Returns a Bool. See `API.h5z_register`.
+"""
+encoder_present(::F) where {F<:Filter} = encoder_present(F)
+encoder_present(::Type{F}) where {F<:Filter} = true
+
+"""
+    decoder_present(::F) where {F<:Filter}
+
+Can the filter decode or decompress the data?
+Defaults to true.
+Returns a Bool.
+See `API.h5z_register`
+"""
+decoder_present(::F) where {F<:Filter} = decoder_present(F)
+decoder_present(::Type{F}) where {F<:Filter} = true
+
+"""
+    filtername(::F) where {F<:Filter}
+
+What is the name of a filter?
+Defaults to "Unnamed Filter"
+Returns a String describing the filter. See `API.h5z_register`
+"""
+filtername(::F) where {F<:Filter} = filtername(F)
+filtername(::Type{F}) where {F<:Filter} = "Unnamed Filter"
+
+"""
+    can_apply_func(::F) where {F<:Filter}
+
+Return a function indicating whether the filter can be applied or `nothing` if no function exists.
+The function signature is `func(dcpl_id::API.hid_t, type_id::API.hid_t, space_id::API.hid_t)`.
+See `API.h5z_register`
+"""
+can_apply_func(::F) where {F<:Filter} = can_apply_func(F)
+can_apply_func(::Type{F}) where {F<:Filter} = nothing
+function can_apply_cfunc(::Type{F}) where {F<:Filter}
+    func = can_apply_func(F)
+    if func === nothing
+        return C_NULL
+    else
+        return @cfunction($func, API.herr_t, (API.hid_t,API.hid_t,API.hid_t))
+    end
+end
+
+"""
+    set_local_func(::F) where {F<:Filter}
+
+Return a function that sets dataset specific parameters or `nothing` if no function exists.
+The function signature is `func(dcpl_id::API.hid_t, type_id::API.hid_t, space_id::API.hid_t)`.
+See `API.h5z_register`
+"""
+set_local_func(::F) where {F<:Filter} = set_local_func(F)
+set_local_func(::Type{F}) where {F<:Filter} = nothing
+function set_local_cfunc(::Type{F}) where {F<:Filter}
+    func = set_local_func(F)
+    if func === nothing
+        return C_NULL
+    else
+        return @cfunction($func, API.herr_t, (API.hid_t,API.hid_t,API.hid_t))
+    end
+end
+
+
+"""
+    filter_func(::F) where {F<:Filter}
+
+Returns a function that performs the actual filtering.
+
+See `API.h5z_register`
+"""
+filter_func(::F) where {F<:Filter} = filter_func(F)
```

Comment:
Unless the functions on instances are actually used, I would remove them and just keep the type ones.

Reply:
Can you be more specific? Do you want to remove `filtername(::F) where F <: Filter` as well?
```julia
julia> HDF5.Filters.filtername(BloscFilter())
ERROR: MethodError: no method matching filtername(::BloscFilter)
Closest candidates are:
  filtername(::HDF5.Filters.UnknownFilter) at C:\Users\kittisopikulm\.julia\dev\HDF5\src\filters\filters.jl:157
  filtername(::Type{HDF5.Filters.UnknownFilter}) at C:\Users\kittisopikulm\.julia\dev\HDF5\src\filters\filters.jl:158
  filtername(::Type{HDF5.Filters.H5Zblosc.BloscFilter}) at C:\Users\kittisopikulm\.julia\dev\HDF5\src\filters\H5Zblosc.jl:139
  ...
Stacktrace:
 [1] top-level scope
   @ REPL[66]:1

julia> HDF5.Filters.filtername(BloscFilter)
"blosc"

julia> HDF5.Filters.filtername(typeof(BloscFilter()))
"blosc"
```

Reply:
I guess unless you need it?

Reply:
I'm conflicted here since the general public API should be invoking this on the instance. It allows for more general implementations of filters. Using `filtername` or `filter_func` on an instance always works at the moment whether it is an `UnknownFilter` or a `BloscFilter` instance. Being able to garner the information statically from a type is the special case here with an important technical detail (a la `@cfunction`).

What I'm thinking about is if a package like JLD2 wanted to query information about the filters to do HDF5-compatible filtering or compression.

Reply:
After thinking about this, I concur. I've removed the methods on the `Filter` instances in c450eba